core/server/master/src/main/java/alluxio/master/file/UfsSyncChecker.java

```diff
@@ -0,0 +1,163 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.master.file;
+
+import alluxio.AlluxioURI;
+import alluxio.exception.FileDoesNotExistException;
+import alluxio.exception.InvalidPathException;
+import alluxio.master.file.meta.Inode;
+import alluxio.master.file.meta.InodeDirectory;
+import alluxio.master.file.meta.MountTable;
+import alluxio.underfs.UnderFileStatus;
+import alluxio.underfs.UnderFileSystem;
+import alluxio.underfs.options.ListOptions;
+import alluxio.util.io.PathUtils;
+
+import com.google.common.base.Preconditions;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * Class to check if the contents of the under storage is in-sync with the master.
+ */
+@NotThreadSafe
+public final class UfsSyncChecker {
+
+  /** UFS directories for which list was called. */
+  private Map<String, UnderFileStatus[]> mListedDirectories;
+
+  /** This manages the file system mount points. */
+  private final MountTable mMountTable;
+
+  /**
+   * Create a new instance of {@link UfsSyncChecker}.
+   *
+   * @param mountTable to resolve path in under storage
+   */
+  public UfsSyncChecker(MountTable mountTable) {
+    mListedDirectories = new HashMap<>();
+    mMountTable = mountTable;
+  }
+  /**
+   * Check if immediate children of directory are in sync with UFS.
+   *
+   * @param inode read-locked directory to check
```

Comment:
it assumes that `inode.getChildren`does not change during execution of this call